internals-js/src/operations.ts

```diff
@@ -1535,6 +1547,53 @@ export class SelectionSet {
     this._selections = mapValues(keyedSelections);
   }
 
+  /**
+   * 
+   * @param namedFragments {NamedFragments}
+   * @param fragmentDefinitionsById 
+   * @returns 
+   */
+  minimizeSelectionSet(
+    namedFragments: NamedFragments = new NamedFragments(),
+    fragmentDefinitionsById: Map<number, [SelectionSet, NamedFragmentDefinition][]> = new Map(),
+  ): [SelectionSet, NamedFragments] {
+    const minimizedSelectionSet = this.lazyMap((selection) => {
+      if (selection.kind === 'FragmentSelection' && selection.element.typeCondition && selection.element.appliedDirectives.length === 0 && selection.selectionSet) {
+        const id = selection.key().length + selection.selectionSet.selections().length;
```

Comment:
for future: we could recursively calculate the selection set sizes to come up with more unique keys but unsure how much we would gain from it

Reply:
also `selection.key().length` might not be the best identifier `... on Foo` has same length as `... on Bar`. Maybe keep it as string instead?

Reply:
[c00429d](https://github.com/apollographql/federation/pull/2958/commits/c00429db96e3e983ef1f8c21b831e931e19a8d1f)